### PR TITLE
Assert pvl group equal

### PR DESCRIPTION
**Pull request by gsn9**
## Description
The test's for checking if AssertPvlGroupEqual were not showing a failure, the problem was that when a failure was detected no failure was being added. To fix this I added the failures after each check to see if a failure had occurred.  

## Related Issue

https://github.com/USGS-Astrogeology/ISIS3/issues/4203
AssertPvlGroupEqual comparison not failing when groups are not equal #4203

## Motivation and Context

The problem was that the test's that compare PvlGroups would always pass even if the group's had different content's. 

## How Has This Been Tested?

Test's have been written in the PvlObjectTest.ccp file, to test my changes I created a new PvlGroup  and various keywords. I then made a copy of the PvlGroup and compared the two with out altering either one. Then I altered the copy and compared them again to see if the changes were caught and that I got a test failure. 

Note: I also added an import for "TestUtilities.h" to be able to call AssertPvlGroupEqual from this file. 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have read and agree to abide by the [Code of Conduct](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/Code-Of-Conduct.md)
- [x] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] I have added myself to the [.zenodo.json](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/.zenodo.json) document.
- [x] I have added any user impacting changes to the [CHANGELOG.md](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CHANGELOG.md) document.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [x] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

### CODE REVIEW

File: isis/tests/TestUtilities.cpp
Comment (Kelvinrr) [inline isis/tests/TestUtilities.cpp]: So does ::testing::AssertionFailure just not work for asserting failures we need to add the `ADD_FAILURE` line?

**Review by Kelvinrr — DISMISSED** [review]

File: isis/tests/TestUtilities.cpp
Comment (gsn9) [reply isis/tests/TestUtilities.cpp]: yeah ::testing::AssertionFailure won't work for asserting a failure with out adding the ``ADD_FAILURE`` line since by default ::testing::AssertionFailure will return True if something has failed.

**Review by acpaquette — CHANGES_REQUESTED** [review]
Two Failing tests on this PR:
```
The following tests FAILED:
	1813 - MroHiriseCube.FunctionalTestHimosDefault (Failed)
	1861 - kaguyatc2isisTest.FunctionalTestKaguyami2isisNir (Failed)
```

**Comment by jessemapel** [comment]
It's very likely those tests were already failing but until this was fixed we didn't see it. Do we want to fix them on here or a different PR?

**Comment by acpaquette** [comment]
@jessemapel @gsn9 I realize now that we may just not be using the `AssertPvlGroupEqual` function correctly. We should be using it like the following example from gtest `EXPECT_PRED_FORMAT2(AssertMutuallyPrime, a, b);` but something along the lines of `EXPECT_PRED_FORMAT2(AssertPvlGroupEqual, args ...)`.

We should convert the uses of `AssertPvlGroupEqual` to the above format rather than adding the `ADD_FAILURE` lines

**Comment by gsn9** [comment]
> e following tests FAILED:
> 	1813 - MroHiriseCu

These test's were failing before I made any changes

File: isis/tests/PvlObjectTests.cpp
Comment (acpaquette) [inline isis/tests/PvlObjectTests.cpp:214]: There should be a different `EXPECT_PRED_FORMAT{num}` where num is the number of arguments. So in this case, we should be able to run something like `EXPECT_PRED_FORMAT4(AssertPvlGroupEqual("Point_ErrorMagnitude"...`

File: isis/tests/PvlObjectTests.cpp
Comment (acpaquette) [inline isis/tests/PvlObjectTests.cpp:222]: Same as previous comment

File: isis/tests/PvlObjectTests.cpp
Comment (acpaquette) [inline isis/tests/PvlObjectTests.cpp:229]: Same as previous comment

**Review by acpaquette — COMMENTED** [review]
Should be the last few changes, ignore the `EXPECT_PRED_FORMAT{num}` comment if we need those to comeback as false.

There are two other files that use this function incorrectly that need to be changed, namely in [FunctionalTestsHimos.cpp](https://github.com/USGS-Astrogeology/ISIS3/blob/aab0e63088046690e6c031881825596c1c2cc380/isis/tests/FunctionalTestsHimos.cpp) on lines 50 and 52 and [FunctionalTestsKaguyami2isis.cpp](https://github.com/USGS-Astrogeology/ISIS3/blob/aab0e63088046690e6c031881825596c1c2cc380/isis/tests/FunctionalTestsKaguyami2isis.cpp) on lines 83 and 161.

**Comment by gsn9** [comment]
> There are two other files that use this function incorrectly that need to be changed, namely in [FunctionalTestsHimos.cpp](https://github.com/USGS-Astrogeology/ISIS3/blob/aab0e63088046690e6c031881825596c1c2cc380/isis/tests/FunctionalTestsHimos.cpp) on lines 50 and 52 and [FunctionalTestsKaguyami2isis.cpp](https://github.com/USGS-Astrogeology/ISIS3/blob/aab0e63088046690e6c031881825596c1c2cc380/isis/tests/FunctionalTestsKaguyami2isis.cpp) on lines 83 and 161.

@acpaquette  I just noticed that for the test's in the [FunctionalTestsHimos.cpp](https://github.com/USGS-Astrogeology/ISIS3/blob/aab0e63088046690e6c031881825596c1c2cc380/isis/tests/FunctionalTestsHimos.cpp) that were currently using `AssertPvlGroupEqual` to test `PvlObject`'s when it should be taking in `PvlGroup`'s, from what I can see it doesn't look like this test is compatible with `PvlObject`'s  and I'm assuming that's why I'm getting a type error when trying to convert the test to use `EXPECT_PRED_FORMAT{num}` . Do we want to modify the test to try and cast both`inputCubeLabel` and `outputCubeLabel` to `PvlGroup`'s? 

error: `could not convert ‘(const char*)"inputCubeLabel"’ from ‘const char*’ to ‘Isis::PvlGroup’
   EXPECT_PRED_FORMAT4(AssertPvlGroupEqual, "InputMapping", "OutputMapping", inputCubeLabel, outputCubeLabel.findGroup("Mapping"));`

**Comment by acpaquette** [comment]
@gsn9 Looks like you're just missing a `findGroup` call on the `inputcCubeLabel`. The function call should look like this:
`EXPECT_PRED_FORMAT4(AssertPvlGroupEqual, "InputMapping", "OutputMapping", inputCubeLabel.findGroup("Mapping"), outputCubeLabel.findGroup("Mapping"))`

**Comment by gsn9** [comment]
@acpaquette  adding the ``findGroup`` call doesn't seem to fix the type error.

Error: ``error: could not convert ‘(const char*)"inputCubeLabel.findGroup(\"Mapping\")"’ from ‘const char*’ to ‘Isis::PvlGroup’
 EXPECT_PRED_FORMAT4(AssertPvlGroupEqual, "InputMapping", "OutputMapping", inputCubeLabel.findGroup("Mapping"), outputCubeLabel.findGroup("Mapping"));``

**Comment by acpaquette** [comment]
@gsn9 I misunderstood the function signature for the `EXPECT_PRED_FORMAT4` the `AssertPvlGroupEqual` function is built on only passing in two pvl groups but the function signature splits out the group and the variable name into seperate arguments. The function call when called with `EXPECT_PRED_FORMAT{num}` should just be used with the PVL groups something like:

`EXPECT_PRED_FORMAT2(AssertPvlGroupEqual, inputCubeLabel.findGroup("Mapping"), outputCubeLabel.findGroup("Mapping"))`

We may have to put those two groups into variable names, then pipe those variables into the function. Something along the lines of this:
```
PvlGroup inputMappingGroup = inputCubeLabel.findGroup("Mapping");
PvlGroup outputMappingGroup = outputCubeLabel.findGroup("Mapping");

EXPECT_PRED_FORMAT2(AssertPvlGroupEqual, inputMappingGroup, outputMappingGroup);
```

**Review by jessemapel — DISMISSED** [review]

**Comment by jessemapel** [comment]
@acpaquette Can you dismiss your change request review?

**Comment by jessemapel** [comment]
I want to wait for CI on this to check the ProcessMapMosaic change doesn't have any unintended changes.

**Review by acpaquette — DISMISSED** [review]

**Comment by acpaquette** [comment]
@jessemapel A few failing tests here due to the `ProcessMapMosaic` change.

Namely:
`isis_unit_test_ProcessMapMosaic`
`tgocassismos_app_test_default`
`MroHiriseCube.FunctionalTestHimosDefault`

File: isis/src/base/objs/ProcessMapMosaic/ProcessMapMosaic.cpp
Comment (jessemapel) [inline isis/src/base/objs/ProcessMapMosaic/ProcessMapMosaic.cpp:550]: We're missing the min/max lat/lon values being passed into this function here. The unit test is failing because it's processing multiple cubes through `SetOutputCube(FileList, CubeAttributeOutput,
      const QString)`. In that function is computes the min/max lat/lon from all input cubes and then passes that to this function. Since we're no longer adding those values, it's only pulling the min/max lat/lon from the first input cube.

**Review by jessemapel — APPROVED** [review]

**Comment by acpaquette** [comment]
To get this fully passing there are two changes to a truthfile (tgocassis.pvl):

In mapping group
UpperLeftCornerX     = -36053.388090957 -->  UpperLeftCornerX     = -36053.388090957 <meters>
UpperLeftCornerY     = -719558.17674986 -->  UpperLeftCornerY     = -719558.17674986 <meters>

In tracking group
SolarLongitude            = 269.128636806 <degrees> --> SolarLongitude            = 269.12863680949 <degrees>

**Pull request merged**
